### PR TITLE
[GEOT-6494] MBStyle extension

### PR DESCRIPTION
**Pull request by jodygarnett**
Reviewing mbstyle codebase, cleaning up reports and checking code coverage.

* https://osgeo-org.atlassian.net/browse/GEOT-6493

Related:

* https://github.com/geotools/geotools/pull/2775 background color
* https://github.com/geotools/geotools/pull/2782 opacity fix

Working on:

* [x] Update pom.xml and move to extension
* [x] Documentation
* [x] Test coverage
* [x] Include BSD license for documentation

## Checklist

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer.

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by aaime** [comment]
@jodygarnett I see this is just a review, which is fine. Could you make sure all PRs against the module are merged before moving it? I have one out, one in the making.

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/layer/CircleMBLayer.java [hunk 1]
```diff
@@ -285,7 +285,7 @@ public Color getCircleStrokeColor() throws MBFormatException {
      * Access circle-stroke-color as literal or function expression, defaults to black.
      *
      * @return The color of the circle stroke.
-     * @throws MBFormatException
+     * @throws MBFormatException JSON provided does conform to specification
```
Comment (aaime) [inline modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/layer/CircleMBLayer.java]: "does not" I suppose?

**Comment by jodygarnett** [comment]
@aaime wrote:

> Could you make sure all PRs against the module are merged before moving it?

Yes I would like to clean up what I can here, and then do a separate PR to move.

**Comment by jodygarnett** [comment]
> @jodygarnett I see this is just a review, which is fine. Could you make sure all PRs against the module are merged before moving it? I have one out, one in the making.

@aaime do you have any additional PRs against this module for me to review?

**Comment by aaime** [comment]
Yes there are two, I pinged you in both already, the background one and the
exponential math one



Il ven 31 gen 2020, 22:53 Jody Garnett <notifications@github.com> ha
scritto:

**Comment by aaime** [comment]
@jodygarnett I don't have time to work on the alpha issue, please go ahead with the review and move, I'll resume later.

File: docs/user/extension/mbstyle/index.rst [hunk 2]
```diff
@@ -43,18 +43,19 @@ Zoom level tuning
...
+   System.setSystemProperty("MBSTYLE_ROOT_TILE_PIXELS", "1024") 
 
 MapBox Types
 ^^^^^^^^^^^^
-  copied from the `MapBox Style Specification <https://www.mapbox.com/mapbox-gl-js/style-spec/>`_
 
-  Color
+Quoting `MapBox Style Specification <https://www.mapbox.com/mapbox-gl-js/style-spec/>`__ :
+
+  ``Color``
```
Comment (jodygarnett) [inline docs/user/extension/mbstyle/index.rst]: @tbarsballe this section has some duplication with `spec/types.rst`, do you know any background to untangle?

**Comment by jodygarnett** [comment]
The integration tests show what I hope is an unrelated failure in GeoServer GeoWebCache (GWC) Module.

File: docs/user/welcome/license.rst [hunk 3]
```diff
@@ -5,12 +5,6 @@ The GeoTools project is a member of the Open Source Geospatial Foundation; the f
 
 As an open source project GeoTools is made available to you using an open source (LGPL) license as described below.
 
-A few select files have additional license:
-
-* Code contributed by government members is often contributed as public domain
-* To facilitate adopting of the GeoTools library the example code used in the
-  user guide is also public domain
-
```
Comment (jodygarnett) [inline docs/user/welcome/license.rst]: Moved this to the respective headings.

File: modules/plugin/pom.xml [hunk 4]
```diff
@@ -59,14 +62,10 @@
     <module>imagemosaic-jdbc</module>
     <module>jdbc</module>
     <module>jp2k</module>
+    <module>mongodb</module>
     <module>property</module>
     <module>shapefile</module>
     <module>svg</module>
-    <module>coverage-multidim</module>
-    <module>geopkg</module>
```
Comment (aaime) [inline modules/plugin/pom.xml]: So this change kicked geopkg out of the build? Ideally a module promotion should just do that, promote the module, all the other improvements should be in a different commit (if not different PR), because they are not related, and are easy to get lost in the movement of so many files (github reports 200+ files changed in this PR). Thankfully I spotted this one at the source...

File: pom.xml [hunk 5]
```diff
@@ -1577,6 +1577,13 @@
         <artifactId>jt-shadedrelief</artifactId>
         <version>${jaiext.version}</version>
       </dependency>
+      
+      <!-- Third-party -->
+      <dependency>
```
Comment (aaime) [inline pom.xml]: Ah, so mbstyle is using json-simple to parse JSON? The library is completely dead, has not seen releases since 2012.
Not holding it as a blocker for the module, but wondering, is there any plan to switch to something supported?

File: docs/user/extension/mbstyle/index.rst (hunk 2 above)
Comment (tbarsballe) [reply docs/user/extension/mbstyle/index.rst]: `spec/*.rst` is a direct copy of the MapBox spect with GeoTools supported version added.
This index page tries to provide a more concise overview / quickstart, and pulls some of the more top-level information from the spec. It could maybe do with some rewording, but should still keep all that info in it.

File: modules/plugin/pom.xml (hunk 4 above)
Comment (jodygarnett) [reply modules/plugin/pom.xml]: Thanks for the catch Andrea, I appreciate it.
I initially tried moving this to plugin, but it did not fit (since it defines new API).

File: pom.xml (hunk 5 above)
Comment (jodygarnett) [reply pom.xml]: I am not aware why this was chosen, just annoyed it was not in the dependency management section.

Didn't we talk this out with GeoMesa folks to select a decent JSON library?

Comment (aaime) [reply pom.xml]: There was a discussion on dependencies for gt-main, and we ended up using jackson for that. But mbstyles is older, and used the same library as the semi-defunct gt-geojson (which I hope is gonna be replaced by the gt-jacksondatastore one day or the other...)

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/layer/CircleMBLayer.java (hunk 1 above)
Comment (jodygarnett) [reply modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/layer/CircleMBLayer.java]: changed to `JSON provided inconsistent with specificaiton`

File: docs/user/extension/mbstyle/index.rst (hunk 2 above)
Comment (jodygarnett) [reply docs/user/extension/mbstyle/index.rst]: okay I removed the duplication, and made this focus on the example snippets

**Comment by jodygarnett** [comment]
Travis is failing on documentation `toctree` use of `caption` directive, must be an older version of sphinx.

**Comment by jodygarnett** [comment]
This was approved by PMC:

* Andrea Aime +1
* Ian Turton +1
* Jody Garnett +1
* Nuno Oliveira +1
* Simone Giannecchini +1
* Torben Barsballe +1

**Comment by jodygarnett** [comment]
There is no need to backport at this time.
